util/osuser/user.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package osuser implements OS user lookup. It's a wrapper around os/user that
+// works on non-cgo builds.
+package osuser
+
+import (
+	"context"
+	"errors"
+	"log"
+	"os"
+	"os/exec"
+	"os/user"
+	"runtime"
+	"strings"
+	"time"
+	"unicode/utf8"
+
+	"tailscale.com/version/distro"
+)
+
+// LookupId is like os/user.LookupId but handles a few edge cases like gokrazy
+// and non-cgo lookups, and returns the user shell.
+func LookupId(uid string) (u *user.User, shell string, err error) {
+	return lookup(uid, user.LookupId)
+}
+
+// Lookup is like os/user.Lookup but handles a few edge cases like gokrazy and
+// non-cgo lookups, and returns the user shell.
+func Lookup(username string) (u *user.User, shell string, err error) {
+	return lookup(username, user.Lookup)
+}
+
+func lookup(usernameOrUID string, lookupStd func(string) (*user.User, error)) (*user.User, string, error) {
+	if runtime.GOOS != "linux" {
+		u, err := lookupStd(usernameOrUID)
+		return u, fallbackShell(), err
+	}
+
+	// No getent on Gokrazy. So hard-code the login shell.
+	if distro.Get() == distro.Gokrazy {
+		shell := "/tmp/serial-busybox/ash"
+		u, err := lookupStd(usernameOrUID)
+		if err != nil {
+			return &user.User{
+				Uid:      "0",
+				Gid:      "0",
+				Username: "root",
+				Name:     "Gokrazy",
+				HomeDir:  "/",
+			}, shell, nil
+		}
+		return u, shell, nil
+	}
+
+	// On Linux, default to using "getent" to look up users so that
+	// even with static tailscaled binaries without cgo (as we distribute),
+	// we can still look up PAM/NSS users which the standard library's
+	// os/user without cgo won't get (because of no libc hooks).
+	// But if "getent" fails, userLookupGetent falls back to the standard
+	// library anyway.
+	return userLookupGetent(usernameOrUID, lookupStd)
+}
+
+func validUsername(usernameOrUID string) bool {
+	maxUid := 32
+	if runtime.GOOS == "linux" {
+		maxUid = 256
+	}
+	if len(usernameOrUID) > maxUid || len(usernameOrUID) == 0 {
+		return false
+	}
+	for _, r := range usernameOrUID {
+		if r < ' ' || r == 0x7f || r == utf8.RuneError { // TODO(bradfitz): more?
+			return false
+		}
+	}
+	return true
+}
+
+func userLookupGetent(usernameOrUID string, lookupStd func(string) (*user.User, error)) (*user.User, string, error) {
```

Comment:
all these "usernameOrUID" sound like these funcs are overloaded and accept either one so I was looking at first where the magic was that heuristically decides whether it's a UID vs a username (and then worrying about security problems with said heuristic) and then I realized what's going on: that user.LookupId also takes a string and you wanted the same wrapper.

Maybe `lookupFuncArg` in general?

But in this case, isn't it always a username and not a UID?

Reply:
I'll need lookup by UID here: https://github.com/tailscale/tailscale/pull/10163/files#r1385655690
As far as ambiguity - `getent` handles both username and UIDs. Can you elaborate on `lookupFuncArg`?

Reply:
> As far as ambiguity - getent handles both username and UIDs. 

Ah, cool. I see that in the manpage now. And we use it for UID lookups now too. Got it.

> Can you elaborate on lookupFuncArg

No need now.